cl/citations/tests.py

```diff
@@ -1085,3 +1093,267 @@ def _print_failed_cases(
+                        id=11,
+                        score=0,
+                    )
+                ],
+            ),
+            (
+                [
+                    DummyParenthetical(
+                        text="The loss of First Amendment freedoms, for even minimal period of time, unquestionably constitutes irreparable injury.",
+                        id=11,
+                        score=0,
+                    ),
+                    DummyParenthetical(
+                        text="The loss of First Amendment freedoms, for even minimal period of time, unquestionably constitutes irreparable injury.",
+                        id=12,
+                        score=0,
+                    ),
+                    DummyParenthetical(
+                        text="The loss of First Amendment freedoms, for even minimal period of time, unquestionably constitutes irreparable injury.",
+                        id=13,
+                        score=0,
+                    ),
+                ],
+                [
+                    DummyParenthetical(
+                        text=" Holding public employees could not be fired because of their politics unless they held “policymaking” or “confidential” positions ",
+                        id=14,
+                        score=0,
+                    ),
+                ],
+            ),
+        ]
+        for i, groups in enumerate(expected_groups):
+            with self.subTest(f"Testing {groups} are grouped correctly.", i=i):
+                flat = sum(groups, start=[])
+                output_groups = get_parenthetical_groups(flat)
+                output_sets = frozenset(
+                    [frozenset(pg.parentheticals) for pg in output_groups]
+                )
+                input_sets = frozenset([frozenset(g) for g in groups])
+                self.assertEquals(
+                    input_sets,
+                    output_sets,
+                    f"Got incorrect result from get_parenthetical_groups for: {groups}",
+                )
+
+    def test_get_representative_parenthetical(self):
+        simgraph = {
+            "0": ["3"],
+            "1": ["2", "3", "7"],
+            "2": ["1"],
+            "3": ["0", "1"],
+            "4": ["5"],
+            "5": ["4"],
+            "6": [],
+            "7": ["1"],
+        }
+
+        parentheticals = [
+            DummyParenthetical(id=0, text="par0", score=1),
+            DummyParenthetical(id=1, text="par1", score=1),
+            DummyParenthetical(id=2, text="par2", score=1),
+            DummyParenthetical(id=3, text="par3", score=1),
+            DummyParenthetical(id=4, text="par4", score=1),
+            DummyParenthetical(id=5, text="par5", score=1),
+            DummyParenthetical(id=6, text="par6", score=1),
+            DummyParenthetical(id=7, text="par7", score=1),
+        ]
+        test_pairs = [
+            ((parentheticals[0:3], simgraph), parentheticals[0]),
```

Comment:
A word here about why parenthetical[0] won would be helpful. My interpretation is that it's because zero is connected to 3, which connects to zero and 1. One connects to 2, 3, 7, and 2 connects to 1, we already counted 3, and 7 connects back to 1 that we already counted.

In other words:

```
┌─────┐
│     │
├───┐ │ ┌─┐
│   │ │ │ │
0 1 2 3 4 5 6 7
  │ │ │       │
  │ │ │       │
  ├─┘ │       │
  │   │       │
  ├───┘       │
  │           │
  └───────────┘
```

But that's hard to see just looking at the data. My comment is pretty not great (but the graph is so-so!). Maybe there's a way to make this more intuitive?

Reply:
Whoa, I'm not even sure "0" is supposed to win there --- let me double check these tests.

Reply:
This behavior isn't a bug - it just looks weird because of how the test is constructed. Remember that `get_representative_parenthetical` only looks at the top 20% (rounded up to the nearest integer) of parentheticals passed in. 20% of 3 rounded up is 1, and so the algorithm only ever looks at the first parenthetical.

> But that's hard to see just looking at the data. My comment is pretty not great (but the graph is so-so!). Maybe there's a way to make this more intuitive?

This is a good thought, but this part of the algorithm is actually much simpler than this: we just rank by the number of immediate neighbors each node has. See above for the explanation as to why this test gives the answer it does.

Reply:
Oh oops --- I definitely knew that while writing the test and forgot in between,  point in case for comments :sweat: